### PR TITLE
Tasks UI

### PR DESCRIPTION
**Pull request by BIRD311**
basic UI so we can actually use the lumpertasks

### CODE REVIEW

File: src/Lumper.UI/ViewModels/Tasks/ChangeTextureTaskViewModel.Item.cs [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+            get => _source;
+            set
+            {
+                if (value != _source && value is not null)
+                {
+                    Remove();
+                    _source = value;
+                    Add(Source, Target);
+
+                    this.RaisePropertyChanged(nameof(Source));
```
Comment (tsa96) [inline src/Lumper.UI/ViewModels/Tasks/ChangeTextureTaskViewModel.Item.cs:40]: From https://www.reactiveui.net/api/reactiveui/ireactiveobjectextensions/63fddd3b: "Leave null to let the runtime set to caller member name.". So you don't need the nameof, just this.RaisePropertyChanged() is fine.

File: src/Lumper.UI/ViewModels/Tasks/ChangeTextureTaskViewModel.Item.cs [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        public string Target
+        {
+            get => _target;
+            set
+            {
+                if (value != _target && value is not null)
+                {
+                    UpdateValue(value);
+                    _target = value;
+                    this.RaisePropertyChanged(nameof(Target));
```
Comment (tsa96) [inline src/Lumper.UI/ViewModels/Tasks/ChangeTextureTaskViewModel.Item.cs:55]: Same as above

File: src/Lumper.UI/ViewModels/Tasks/TasksViewModel.cs [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+        BspFile = bsp;
+        TaskTypes = new()
+        {
+            new TaskMenuItem(this, typeof(ChangeTextureTask)),
+            new TaskMenuItem(this, typeof(StripperTask)),
+            new TaskMenuItem(this, typeof(CompressionTask)),
+            new TaskMenuItem(this, typeof(RunExternalToolTask)),
+        };
+        this.WhenAnyPropertyChanged(nameof(SelectedTask))
+         .Subscribe(x => OnSelectedTaskChanged());
```
Comment (tsa96) [inline src/Lumper.UI/ViewModels/Tasks/TasksViewModel.cs:33]: Best to use `_` for unused params

File: src/Lumper.UI/ViewModels/Tasks/TasksViewModel.cs [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+        if (SelectedTask is null || Tasks.Count <= 1)
+            return;
+        int idx = Tasks.IndexOf(SelectedTask);
+        int newIdx = idx + 1;
+        if (newIdx >= Tasks.Count)
+            newIdx = 0;
+        Tasks.Move(idx, newIdx);
+        SelectedTask = null;
+        SelectedTask = Tasks[newIdx];
+    }
```
Comment (tsa96) [inline src/Lumper.UI/ViewModels/Tasks/TasksViewModel.cs:111]: Can we just generalise these two to `MoveSelectedTask(int dir)`

File: src/Lumper.UI/ViewModels/Tasks/ChangeTextureTaskViewModel.Item.cs [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+            }
+        }
+
+        protected string _target = "";
+        public string Target
+        {
+            get => _target;
+            set
+            {
+                if (value != _target && value is not null)
```
[the remark this answers is not in the document]
Comment (tsa96) [reply src/Lumper.UI/ViewModels/Tasks/ChangeTextureTaskViewModel.Item.cs:51]: also early return

File: src/Lumper.UI/Converters/TaskStatusConverter.cs [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+                case TaskStatus.Failed:
+                    return "⚠";
+                case TaskStatus.Unknown:
+                default:
+                    return " ";
+            }
+        }
+        return new BindingNotification(
+            new ArgumentOutOfRangeException(nameof(value)),
+            BindingErrorType.DataValidationError);
```
Comment (tsa96) [inline src/Lumper.UI/Converters/TaskStatusConverter.cs:41]: Just return the notif if `value is not TaskStatus status`

File: src/Lumper.UI/ViewModels/Tasks/TaskViewModel.cs [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    public bool _isRunning;
+    public bool IsRunning
+    {
+        get => _isRunning;
+        set => this.RaiseAndSetIfChanged(ref _isRunning, value);
+    }
+
+    public bool Run(BspFile bsp)
+    {
+        bool ret = true;
```
Comment (tsa96) [inline src/Lumper.UI/ViewModels/Tasks/TaskViewModel.cs:56]: Give clearer name like `success`, `wasSuccessful`, w/e

File: src/Lumper.UI/Views/Tasks/TasksView.axaml.cs [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    }
+
+    private void InitializeComponent()
+    {
+        AvaloniaXamlLoader.Load(this);
+    }
+
+    private static IReadOnlyList<FilePickerFileType> GenerateJsonFileFilter()
+    {
+        var bspFilter = new FilePickerFileType("Json files");
```
Comment (tsa96) [inline src/Lumper.UI/Views/Tasks/TasksView.axaml.cs:40]: "JSON" not "Json"

File: src/Lumper.UI/Views/Tasks/TasksView.axaml.cs [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+    }
+
+    public async void OnLoadClick(object sender, RoutedEventArgs e)
+    {
+        if (Desktop.MainWindow is null)
+            return;
+
+        var dialog = new FilePickerOpenOptions();
+        dialog.AllowMultiple = false;
+        dialog.Title = "Pick BSP file";
```
Comment (tsa96) [inline src/Lumper.UI/Views/Tasks/TasksView.axaml.cs:55]: Shouldn't this be "Load task file"?

File: src/Lumper.UI/Views/Tasks/TasksView.axaml.cs [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+            return;
+        if (DataContext is TasksViewModel vm)
+            vm.Load(await result[0].OpenReadAsync());
+    }
+    public async void OnSaveClick(object sender, RoutedEventArgs e)
+    {
+        if (Desktop.MainWindow is null)
+            return;
+        var dialog = new FilePickerSaveOptions();
+        dialog.Title = "Pick BSP file";
```
Comment (tsa96) [inline src/Lumper.UI/Views/Tasks/TasksView.axaml.cs:68]: Ditto above

File: src/Lumper.UI/Views/Tasks/TasksView.axaml [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+             mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
+             x:Class="Lumper.UI.Views.Tasks.TasksView">
+    <UserControl.Resources>
+        <converters:TaskStatusConverter x:Key="taskStatus" />
+        <converters:PercentConverter x:Key="percent" />
+    </UserControl.Resources>
+    <Grid RowDefinitions="Auto, *, Auto" ColumnDefinitions="*,2*">
+        <DockPanel Grid.Row="0" Grid.Column="0">
+            <Button DockPanel.Dock="Left" Click="OnLoadClick" IsEnabled="{Binding !IsRunning}">Load</Button>
+            <Button DockPanel.Dock="Left" Click="OnSaveClick" IsEnabled="{Binding !IsRunning}">Save</Button>
```
Comment (tsa96) [inline src/Lumper.UI/Views/Tasks/TasksView.axaml:17]: Let's make it clearer what this is doing. We have horizontal space free, let's do "Load Tasks"/"Save Tasks"

**Review by tsa96 — CHANGES_REQUESTED** [review]
Some comments on UI stuff. Styling doesn't matter for now but we should still have the layout in a decent shape.

- Up, Down ,Delete button should be on same row as Load, Save stuff
- The RMB to bring up Add... CM is really unintuitive, even for dev-facing tool this feels bad. I'd do "Add" button next to Up, Down, Delete
- Sidebar should be resizable like the node list in the browser view
Untitled_-_Tom_-_Obsidian_v1 1 1609,04,23,152343

**Comment by BIRD311** [comment]
> Some comments on UI stuff. Styling doesn't matter for now but we should still have the layout in a decent shape.
> 
>     * Up, Down ,Delete button should be on same row as Load, Save stuff
> 
>     * The RMB to bring up Add... CM is really unintuitive, even for dev-facing tool this feels bad. I'd do "Add" button next to Up, Down, Delete
> 
>     * Sidebar should be resizable like the node list in the browser view
>       Untitled_-Tom-_Obsidian_v1 1 1609,04,23,152343

image
On your resolution maybe. This is usable but if I move the buttons to the top row it gets cut off. In the future I would like to get rid if the Up/Down and move it to a context menu on the menu item. Drag and drop to change the order would also be nice. The Delete can also go to the context menu. The Add context menu should probably not be a context menu at all. For now this was just an easy way to do and I don't want to spend to much time on this.

I can move the Up/Down/Delete row below the Load/Save so its closer or try to get it to wrap around to a new line if the window gets to small, but like I said, there should only be 3 buttons (Load, Save and Add)

File: src/Lumper.UI/ViewModels/Tasks/ChangeTextureTaskViewModel.Item.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (BIRD311) [reply src/Lumper.UI/ViewModels/Tasks/ChangeTextureTaskViewModel.Item.cs:51]: strings can be null

File: src/Lumper.UI/ViewModels/Tasks/TasksViewModel.cs (hunk 3 above)
Comment (BIRD311) [reply src/Lumper.UI/ViewModels/Tasks/TasksViewModel.cs:33]: I don't get it .. whats unused?

File: src/Lumper.UI/Converters/TaskStatusConverter.cs (hunk 6 above)
Comment (BIRD311) [reply src/Lumper.UI/Converters/TaskStatusConverter.cs:41]: if I return early I have to cast again .. I like this more .. or what do you mean? `is *not* something *value*` doesn't make sense, does it?

File: src/Lumper.UI/ViewModels/Tasks/TasksViewModel.cs (hunk 3 above)
Comment (tsa96) [reply src/Lumper.UI/ViewModels/Tasks/TasksViewModel.cs:33]: `x`

File: src/Lumper.UI/ViewModels/Tasks/ChangeTextureTaskViewModel.cs [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+        }
+    }
+
+    public void AddRegexItem()
+    {
+        if (Task is ChangeTextureTask c)
+        {
+            var tmp = new ChangeTextureRegexItem(this, c);
+            if (tmp.Add())
+                RegexItems.Add(tmp);
```
Comment (tsa96) [inline src/Lumper.UI/ViewModels/Tasks/ChangeTextureTaskViewModel.cs:47]: Oh sorry, shouldn've brought up the above thing with renaming `tmp` here as well

File: src/Lumper.UI/Converters/TaskStatusConverter.cs (hunk 6 above)
Comment (tsa96) [reply src/Lumper.UI/Converters/TaskStatusConverter.cs:41]: Try it, compiler makes the inference fine! i.e.
```c#
if (value is not TaskStatus status)
    return new BindingNotification(
        new ArgumentOutOfRangeException(nameof(value)),
        BindingErrorType.DataValidationError);
switch (status)
{
    case TaskStatus.Waiting:
        return "|";
    case TaskStatus.Running:
        return "➠";
    case TaskStatus.Success:
        return "✓";
    case TaskStatus.Failed:
        return "⚠";
    case TaskStatus.Unknown:
    default:
        return " ";
}
```

File: src/Lumper.UI/ViewModels/Tasks/ChangeTextureTaskViewModel.Item.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (tsa96) [reply src/Lumper.UI/ViewModels/Tasks/ChangeTextureTaskViewModel.Item.cs:52]: We have nullable reference types on, and `ChangeTextureReplaceItem` ctor takes KV-pair of non-nullable strings. If value can be null, should be KeyValuePair<string, string?>`

**Review by tsa96 — CHANGES_REQUESTED** [review]

File: src/Lumper.UI/ViewModels/Tasks/ChangeTextureTaskViewModel.RegexItem.cs [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+            }
+        }
+
+        public override bool Add(string? source = null,
+                                 string? target = null)
+        {
+            if (source == null)
+                source = Source;
+            if (target == null)
+                target = Target;
```
Comment (tsa96) [inline src/Lumper.UI/ViewModels/Tasks/ChangeTextureTaskViewModel.RegexItem.cs:32]: Crap, sorry, missed the same thing with `??=` here.

**Review by tsa96 — CHANGES_REQUESTED** [review]

File: src/Lumper.UI/ViewModels/Tasks/ChangeTextureTaskViewModel.Item.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (BIRD311) [reply src/Lumper.UI/ViewModels/Tasks/ChangeTextureTaskViewModel.Item.cs:52]: I'm just saying that `string s = null;` works, so `Target = null;` or `Source = null;` works .. I don't want `null` there so check for it .. the only reason I made some of the strings explicitly nullable is to get rid of the warnings you get when doing `string s = null;`

[the remark this answers is not in the document]
Comment (tsa96) [reply src/Lumper.UI/ViewModels/Tasks/ChangeTextureTaskViewModel.Item.cs:52]: Shouldn't you make the value in the kv pair string? then? In which case doesn't the compiler give you different NRT warnings?

**Review by tsa96 — APPROVED** [review]

**Pull request merged**